### PR TITLE
core: handle constraints in infraExplorer instead of pathfinding for stdcm

### PR DESCRIPTION
**Pull request by SarahBellaha**
First part of #6335

### CODE REVIEW

File: core/src/main/kotlin/fr/sncf/osrd/stdcm/infra_exploration/InfraExplorer.kt
Comment (eckter) [inline core/src/main/kotlin/fr/sncf/osrd/stdcm/infra_exploration/InfraExplorer.kt]:
```suggestion
                if (constraints.apply(block).isNotEmpty()) return false
```

File: core/src/main/kotlin/fr/sncf/osrd/stdcm/infra_exploration/InfraExplorerWithEnvelope.kt
Comment (eckter) [inline core/src/main/kotlin/fr/sncf/osrd/stdcm/infra_exploration/InfraExplorerWithEnvelope.kt]: This *could* be a generic `EdgeToRange` instead of a `ConstraintCombiner` (though the name isn't ideal and could use an alias to `PathfindingConstraint`). In the tests, you wouldn't need to instantiate a combiner of a single constraint, you would directly input the electrification constraint.

But it's not really important, it's fine as it is

File: core/src/main/kotlin/fr/sncf/osrd/stdcm/graph/STDCMPathfinding.kt
Comment (eckter) [inline core/src/main/kotlin/fr/sncf/osrd/stdcm/graph/STDCMPathfinding.kt]: I'd prefer if the constraints were given as a parameter to this method, with the constraint instantiation being moved to a single method.

In fact said method could be shared with the block pathfinding, we (should) instantiate the same constraints in the same way.

**Review by eckter — COMMENTED** [review]
Very nice job, I have a few minor comments but LGTM for the most part

File: core/src/main/kotlin/fr/sncf/osrd/stdcm/graph/STDCMPathfinding.kt
Comment (shenriotpro) [reply core/src/main/kotlin/fr/sncf/osrd/stdcm/graph/STDCMPathfinding.kt]: for context https://github.com/osrd-project/osrd/blob/dev/core/src/main/kotlin/fr/sncf/osrd/api/pathfinding/PathfindingBlocksEndpoint.kt#L215

**Review by Erashin — COMMENTED** [review]
Lgtm as well, modulo Eloi's comments which I agree with. Not validating because of the freeze, but it's fine for me :)

**Review by eckter — APPROVED** [review]

**Pull request merged**
